core/src/test/scala/org/apache/spark/rdd/RDDSuite.scala

```diff
@@ -1255,6 +1256,37 @@ class RDDSuite extends SparkFunSuite with SharedSparkContext with Eventually {
     assert(numPartsPerLocation(locations(1)) > 0.4 * numCoalescedPartitions)
   }
 
+  test("SPARK-40211: customize initialNumPartitions for take") {
+    val totalElements = 100
+    val numToTake = 50
+    val rdd = sc.parallelize(0 to totalElements, totalElements)
+    var jobCount = 0
```

Comment:
I'm wondering whether this should be an `AtomicInteger`: the listener invocations happen in a different thread from the main test runner thread and I'm not sure how visibility of writes is guaranteed (it might be happening implicitly via some synchronization somewhere else, but just to be safe we might want to rely on explicit synchronization or atomic operations.

Alternatively, I think we could do something like

```scala
val jobCountListener = new SparkListener {
  private var count = 0
  def getCount: Int = synchronized { count } 
  def reset(): Unit =  synchronized { count = 0}
  override def onJobStart(jobStart: SparkListenerJobStart): Unit = synchronized {
    count += 1
  }
}
sc.addSparkListener(jobCountListener)

...

assert(jobCountListener.getCount > 1)
jobCountListener.reset()
```
```